File: ponio/demos/exp.cpp
// Copyright 2022 PONIO TEAM. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

#include <filesystem>
#include <iostream>
#include <tuple>

#include <ponio/observer.hpp>
#include <ponio/runge_kutta.hpp>
#include <ponio/solver.hpp>

// solve $\dot{u} = u$ with $u(t=0) = 1$, and $t\in[0,2]$.

int
main( int, char** )
{
    std::string const dirname = "exp_data";
    auto filename             = std::filesystem::path( dirname ) / "exp.dat";
    observer::file_observer fobs( filename );

    auto identity = []( double, double u )
    {
        return u;
    };
    double const x0 = 1.0;
    double const dt = 0.1;

    ponio::solve( identity, ponio::runge_kutta::rk_nssp_21(), x0, { 0., 2.0 }, dt, fobs );

    return 0;
}

File: ponio/examples/exp.py
# Copyright 2022 PONIO TEAM. All rights reserved.
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file.

#!/usr/bin/env python

import subprocess
import os
import argparse

import matplotlib.pyplot as plt
import numpy as np

name = "exp"
data_dir = name+"_data"
img_dir = data_dir

parser = argparse.ArgumentParser(
    description=f"Compile, launch and plot results of example `{name}`")
parser.add_argument('--only-save', action='store_true',
                    help="Just save output in img directory")

arguments = parser.parse_args()

if arguments.only_save:
    img_dir = os.path.join("img", name)
    os.makedirs(img_dir, exist_ok=True)


make = subprocess.Popen(["make", name])
make.wait()

args = [os.path.join(".", name)]
process = subprocess.Popen(args)
process.wait()

data = np.loadtxt(os.path.join(data_dir, "exp.dat"))

plt.plot(data[:, 0], np.exp(data[:, 0]), "--", label="exact solution")
plt.plot(data[:, 0], data[:, 1], "-+", label="solution with RK NSSP (2,1)")

plt.legend()

plt.savefig(os.path.join(img_dir, "01.png"))

if not arguments.only_save:
    plt.title("$\\dot{u} = u$, $u(t=0)=1$")
    plt.show()
